clang/lib/Lex/PPDirectives.cpp

```diff
@@ -1995,13 +1995,31 @@ void Preprocessor::HandleIncludeDirective(SourceLocation HashLoc,
     auto Include = CActions->handleIncludeDirective(
         *this, IncludePos, CurLexer->getSourceLocation());
 
+    SmallString<128> FilenameBuffer;
+    StringRef Filename = getSpelling(FilenameTok, FilenameBuffer);
+    SourceLocation CharEnd = FilenameTok.getEndLoc();
+    CharSourceRange FilenameRange =
+        CharSourceRange::getCharRange(FilenameTok.getLocation(), CharEnd);
+    bool isAngled =
+        GetIncludeFilenameSpelling(FilenameTok.getLocation(), Filename);
+    SrcMgr::CharacteristicKind FileCharacter =
+        SourceMgr.getFileCharacteristic(FilenameTok.getLocation());
+
     auto HandleIncludeFile = [&](const PPCachedActions::IncludeFile *File) {
       const FileEntry *FE = SourceMgr.getFileEntryForID(File->FID);
       bool IsImport =
           IncludeTok.getIdentifierInfo()->getPPKeywordID() == tok::pp_import;
       if (FE && IsImport) {
         HeaderInfo.getFileInfo(FE).isImport = true;
       }
+      if (Callbacks && !HashLoc.isInvalid()) {
+        OptionalFileEntryRef FER = SourceMgr.getFileEntryRefForID(File->FID);
+        Callbacks->InclusionDirective(
+            HashLoc, IncludeTok, Filename, isAngled, FilenameRange, FER,
+            /*SearchPath=*/"", /*RelativePath=*/"", /*SuggestedModule=*/nullptr,
+            /*ModuleImported=*/false, FileCharacter);
```

Comment:
In the case of `PPCachedActions::SpuriousImport` (see below), I think we might have a module import that should be passed in here.  CC @jansvoboda11

Reply:
Won't the new change make the callback twice now for the spurious import case? I'm not understanding this

Reply:
@jansvoboda11 can you clarify what callback would be expected here for a spurious import? I would think we'd get a single callback, but maybe I'm misunderstanding.

Reply:
I don't know off the top of my head. Let's try replicate whatever Clang does for implicit modules in this situation.

Reply:
Looking at implicit build, the inclusion only gets a textual header callback. Update to match that behavior.